diff --git a/Makefile b/Makefile
index e279213..4bb7ab5 100644
--- a/Makefile
+++ b/Makefile
@@ -14,6 +14,10 @@ LIB = \
 	eprintf.o \
 	readpassphrase.o
 
+ifeq ($(shell uname -s), Darwin)
+	LIB += osx.o
+endif
+
 SRC = \
 	ratox.c
 
diff --git a/osx.c b/osx.c
new file mode 100644
index 0000000..e29bdc6
--- /dev/null
+++ b/osx.c
@@ -0,0 +1,115 @@
+/* See LICENSE file for copyright and license details. */
+#include <fcntl.h>
+#include <mach/mach_time.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "osx.h"
+
+int
+openat(int dirfd, const char *pathname, int flags, ...)
+{
+	va_list ap;
+	int fd, mode;
+	char *cwd;
+
+	va_start(ap, flags);
+	cwd = getcwd(NULL, 0);
+
+	if (fchdir(dirfd) < 0)
+		return -1;
+
+	if (flags & O_CREAT) {
+		mode = va_arg(ap, int);
+		fd = open(pathname, flags, mode);
+	} else {
+		fd = open(pathname, flags);
+	}
+
+	if (chdir(cwd) < 0)
+		return -1;
+
+	free(cwd);
+	return fd;
+}
+
+int
+mkdirat(int dirfd, const char *pathname, mode_t mode)
+{
+	int rc;
+	char *cwd = getcwd(NULL, 0);
+
+	if (fchdir(dirfd) < 0)
+		return -1;
+
+	rc = mkdir(pathname, mode);
+
+	if (chdir(cwd) < 0)
+		return -1;
+
+	free(cwd);
+	return rc;
+}
+
+int
+mkfifoat(int dirfd, const char *pathname, mode_t mode)
+{
+	int rc;
+	char *cwd = getcwd(NULL, 0);
+
+	if (fchdir(dirfd) < 0)
+		return -1;
+
+	rc = mkfifo(pathname, mode);
+
+	if (chdir(cwd) < 0)
+		return -1;
+
+	free(cwd);
+	return rc;
+}
+
+int
+unlinkat(int dirfd, const char *pathname, int flags)
+{
+	int rc;
+	char *cwd = getcwd(NULL, 0);
+
+	if (fchdir(dirfd) < 0)
+		return -1;
+
+	if (flags & AT_REMOVEDIR)
+		rc = rmdir(pathname);
+	else
+		rc = unlink(pathname);
+
+	if (chdir(cwd) < 0)
+		return -1;
+
+	free(cwd);
+	return rc;
+}
+
+int
+clock_gettime(int clk_id, struct timespec *ts)
+{
+	static mach_timebase_info_data_t timebase_info;
+	uint64_t mach_time, mach_time_nano;
+
+	if (clk_id != CLOCK_MONOTONIC)
+		return -1;
+
+	// initialize timebase_info if necessary
+	if (timebase_info.denom == 0)
+		mach_timebase_info(&timebase_info);
+
+	mach_time = mach_absolute_time();
+	mach_time_nano = mach_time * timebase_info.numer / timebase_info.denom;
+
+	ts->tv_sec = mach_time_nano / 1E9;
+	ts->tv_nsec = mach_time_nano;
+
+	return 0;
+}
diff --git a/osx.h b/osx.h
new file mode 100644
index 0000000..4a9935f
--- /dev/null
+++ b/osx.h
@@ -0,0 +1,9 @@
+/* See LICENSE file for copyright and license details. */
+#define AT_REMOVEDIR    0x200
+#define CLOCK_MONOTONIC 1
+
+int openat(int, const char *, int, ...);
+int mkdirat(int, const char *, mode_t);
+int mkfifoat(int, const char *, mode_t);
+int unlinkat(int, const char *, int);
+int clock_gettime(int, struct timespec *);
diff --git a/ratox.c b/ratox.c
index cf4d38c..fd053c5 100644
--- a/ratox.c
+++ b/ratox.c
@@ -22,6 +22,9 @@
 #include <tox/toxencryptsave.h>
 
 #include "arg.h"
+#ifdef __APPLE__
+#include "osx.h"
+#endif
 #include "queue.h"
 #include "readpassphrase.h"
 #include "util.h"
